lib/Conversion/AffineToPipeline/AffineToPipeline.cpp

```diff
@@ -397,48 +431,101 @@ LogicalResult AffineToPipeline::createPipelinePipeline(
 
   // Maintain mappings of values in the loop body and results of stages,
   // initially populated with the iter args.
-  BlockAndValueMapping valueMap;
-  for (size_t i = 0; i < iterArgs.size(); ++i)
-    valueMap.map(forOp.getBody()->getArgument(i),
-                 pipeline.getStagesBlock().getArgument(i));
+  BlockAndValueMapping iterValueMap;
+  for (size_t i = 0; i < forOp.getBody()->getNumArguments(); ++i)
+    iterValueMap.map(forOp.getBody()->getArgument(i),
+                     pipeline.getStagesBlock().getArgument(i));
 
   // Create the stages.
   Block &stagesBlock = pipeline.getStagesBlock();
   builder.setInsertionPointToStart(&stagesBlock);
 
   // Iterate in order of the start times.
   SmallVector<unsigned> startTimes;
-  for (auto group : startGroups)
+  for (const auto &group : startGroups)
     startTimes.push_back(group.first);
   llvm::sort(startTimes);
 
   DominanceInfo dom(getOperation());
+  SmallVector<DenseSet<Value>> registerValues;
+  SmallVector<SmallVector<mlir::Type>> registerTypes;
+  SmallVector<BlockAndValueMapping> valueMaps;
+  DenseMap<Operation *, std::pair<unsigned, unsigned>> pipeTimes;
   for (auto startTime : startTimes) {
     auto group = startGroups[startTime];
-    OpBuilder::InsertionGuard g(builder);
 
     // Collect the return types for this stage. Operations whose results are not
     // used within this stage are returned.
     auto isLoopTerminator = [forOp](Operation *op) {
       return isa<AffineYieldOp>(op) && op->getParentOp() == forOp;
     };
-    SmallVector<Type> stageTypes;
-    DenseSet<Operation *> opsWithReturns;
+
     for (auto *op : group) {
+      if (op->getUsers().empty())
+        continue;
+      unsigned pipeStartTime = -1;
+      unsigned pipeEndTime = 0;
       for (auto *user : op->getUsers()) {
         if (*problem.getStartTime(user) > startTime || isLoopTerminator(user)) {
-          opsWithReturns.insert(op);
-          stageTypes.append(op->getResultTypes().begin(),
-                            op->getResultTypes().end());
+          pipeEndTime = std::max(pipeEndTime, *problem.getStartTime(user));
+          pipeStartTime = std::min(pipeStartTime, *problem.getStartTime(user));
         }
       }
+      // Skip iter if the users are in the same stage
+      if (pipeStartTime > pipeEndTime)
+        continue;
+
+      pipeTimes[op] = std::pair(pipeStartTime, pipeEndTime);
+
+      // Add register stages for each time slice
+      for (unsigned i = registerValues.size(); i <= pipeEndTime; ++i)
+        registerValues.push_back(DenseSet<Value>());
+
+      for (auto result : op->getResults())
+        registerValues.data()[startTime].insert(result);
```

Comment:
I think we usually avoid indexing into `data()`. If I understand how you're managing `registerValues`, it should be safe to index `registerValues` directly.